### ISSUE TITLE
Temporal Circuits: Mapping Time Horizon Features with Attribution Graphs and SAEs

### ISSUE THREAD

**Issue by justinshenk** [justinshenk/temporal-awareness#23]
# Temporal Circuits: Mapping Time Horizon Features with Attribution Graphs and SAEs

## Overview

Use mechanistic interpretability techniques to reverse-engineer how GPT-2 (and larger models) represent and process temporal information. This extends the probe-based findings by identifying the actual circuits responsible for temporal reasoning.

## Motivation

Current work shows:
- Linear probes detect temporal scope at Layer 8 with 97.1% accuracy
- Steering vectors can shift generation toward short/long-term outputs
- BUT: We don't know *which circuits* encode this information

This experiment maps the temporal circuits using attribution graphs (like your [factual recall circuits work](https://github.com/samjiawng/factual-recall-circuits)) and SAEs (like your Llama 3.1 model-diffing work).

---

## Research Questions

1. **Which attention heads attend to temporal cues?**
2. **Are there dedicated "temporal neurons" or is it distributed?**
3. **Can we find a minimal circuit for temporal classification?**
4. **Do SAE features cleanly separate short vs long-term?**

---

### Phase 1: Attribution Patching for Temporal Decisions

Use activation patching to identify components critical for temporal classification:

```python
# Pseudo-code using TransformerLens
import transformer_lens as tl

model = tl.HookedTransformer.from_pretrained("gpt-2")

# Clean run: "I will plan for retirement" (long-term)
# Corrupt run: "I will eat lunch now" (short-term)

def temporal_patching_experiment():
    """Identify which components change temporal classification."""
    
    # Metric: probe score at Layer 8
    def temporal_metric(logits, cache):
        layer_8_acts = cache["blocks.8.hook_resid_post"]
        return probe.predict_proba(layer_8_acts)[:, 1]  # P(long-term)
    
    # Patch each attention head
    for layer in range(12):
        for head in range(12):
            patched_score = run_with_patch(
                clean="I will plan for retirement",
                corrupt="I will eat lunch now",
                patch_location=f"blocks.{layer}.attn.hook_z",
                head_idx=head
            )
            attribution[layer, head] = patched_score - clean_score
```

**Expected output**: Attribution matrix showing which heads are causally important for temporal classification.

### Phase 2: SAE Feature Analysis

Apply Sparse Autoencoders to find interpretable temporal features:

```python
# Using existing SAE from Neuronpedia or train custom
sae = load_sae("gpt2-layer-8")

# Collect activations on temporal prompts
short_term_acts = collect_activations(SHORT_TERM_PROMPTS)
long_term_acts = collect_activations(LONG_TERM_PROMPTS)

# Find differentially active features
for feature_idx in range(sae.n_features):
    short_activation = sae.encode(short_term_acts)[feature_idx].mean()
    long_activation = sae.encode(long_term_acts)[feature_idx].mean()
    
    if abs(short_activation - long_activation) > threshold:
        temporal_features.append({
            'idx': feature_idx,
            'direction': 'short' if short_activation > long_activation else 'long',
            'diff': abs(short_activation - long_activation)
        })
```

**Expected output**: List of SAE features that encode temporal information, with interpretable descriptions.

### Phase 3: Circuit Mapping

Combine Phases 1-2 to map the full temporal circuit:

```
Input → [Temporal Keyword Detection Heads] → [Integration MLPs] → [Layer 8 Representation] → Output
        L1H3, L2H7 (attend to "now", "future")   L4, L5 (combine cues)   (probe location)
```

Create Neuronpedia-style visualization of the temporal circuit.

---

## Connection to Existing Work

| Prior Work | Extension Here |
|------------|----------------|
| [Factual Recall Circuits](https://github.com/samjiawng/factual-recall-circuits) | Apply same methodology to temporal reasoning |
| [Refusal Directions](https://github.com/samjiawng/refusal-directions) | Compare temporal steering vectors to refusal vectors |
| [Truth Probe Generalization](https://github.com/samjiawng/truth-probe-generalization) | Test temporal probe robustness across domains |
| SAE model-diffing | Find temporal features that differ across model sizes |

---

## Technical Requirements

- TransformerLens for circuit analysis
- SAE toolkit (Neuronpedia, SAELens)
- Existing temporal probe (Layer 8, 97.1% accuracy)
- Temporal prompts dataset (`data/raw/token_temporal_annotations.json`)

---

## Deliverables

1. **Attribution matrix**: Which heads/MLPs are causal for temporal classification
2. **SAE feature catalog**: Interpretable temporal features with descriptions
3. **Circuit diagram**: Visualization of temporal processing pathway
4. **Neuronpedia integration**: Upload features for community exploration
5. **Robustness analysis**: How do circuits differ across domains (finance, health, etc.)?

---

## Implementation Plan

- [ ] Set up TransformerLens pipeline with temporal probe metric
- [ ] Run attention head attribution patching
- [ ] Run MLP attribution patching
- [ ] Load/train SAE for Layer 8
- [ ] Identify top-k temporal features in SAE
- [ ] Manually interpret top features
- [ ] Map circuit diagram
- [ ] Test circuit across OOD temporal prompts
- [ ] Document failure modes
- [ ] Create Neuronpedia visualization

---

## Success Criteria

| Metric | Target |
|--------|--------|
| Identified causal heads | ≥3 heads with attribution >0.1 |
| SAE temporal features | ≥10 clearly interpretable features |
| Circuit faithfulness | Ablating circuit drops accuracy >20% |
| OOD robustness | Circuit works across 3+ domains |

---

## Related Issues

- Issue #4: Synthetic temporal neurons via SAEs
- Issue #19: Token-level temporal scoring
- Issue #20: Temporal logit steering

## References

- Conmy et al. (2023). Towards Automated Circuit Discovery
- Marks et al. (2024). Sparse Feature Circuits
- Templeton et al. (2024). Scaling Monosemanticity (Anthropic)
- [Neuronpedia Attribution Graphs](https://www.neuronpedia.org/gemma-2-2b/graph)
